Send current project scope & CI status headers on request

iarna opened the pull request:
TODO:
- [x] Document CI header behavior
- [x] Document Scope header behavior
- [x] Swap out WIP npm-registry-client with a release that has these patches merged.

iarna committed:
npm: Send Npm-Scope header along with requests to registry

Npm-Scope is set to the scope of the current top level project.

Credit: @iarna
PR-URL: https://github.com/npm/npm/pull/14129
Reviewed-By: @othiym23

iarna committed:
test: Add test to ensure CI header is being sent when env is set

Credit: @iarna
PR-URL: https://github.com/npm/npm/pull/14129
Reviewed-By: @othiym23

othiym23 reviewed (commented):
Looks good so far!

iarna committed:
doc: Document headers we add to registry requests

Credit: @iarna
PR-URL: https://github.com/npm/npm/pull/14129
Reviewed-By: @ashleygwilliams

ashleygwilliams commented on doc/misc/npm-registry.md:
```diff
@@ -21,6 +21,21 @@ The registry URL used is determined by the scope of the package (see
 
+## Does npm send any information about me back to the registry?
+
+When making requests of the registry npm adds two headers with information
```
I'd start this with the word Yes.

iarna replied:
Oh... of course! =D

ashleygwilliams commented on doc/misc/npm-registry.md:
```diff
@@ -21,6 +21,21 @@ The registry URL used is determined by the scope of the package (see
+  continous integration environment, "false" otherwise. This is detected by
+  looking for the following environment variables: `CI`, `TDDIUM`,
+  `JENKINS_URL`, `bamboo.buildKey`. This is used to gather better metrics on
+  how npm is used by humans, versus build farms.
```
might say bots instead of build farm here

iarna replied:
Huh, I would think of bots as something entirely different than CI…

ashleygwilliams commented on doc/misc/npm-registry.md:
```diff
@@ -21,6 +21,21 @@ The registry URL used is determined by the scope of the package (see
 supplied by the `registry` config parameter.  See `npm-config(1)`,
 `npmrc(5)`, and `npm-config(7)` for more on managing npm's configuration.
 
+## Does npm send any information about me back to the registry?
```
i'd also mention if the data sent is tied to username or package. i.e. is it "anonymized", not taken into account with any other data. also is the scope and the ci data going to be analyzed together or kept separate. people might worry about rate limiting.

iarna replied:
I think I can kind of see where your going with this but I'm at a loss as to what to do exactly?

"data sent is tied to username or package" I don't really know what you mean by "tied"? If you're authed then your authentication info can be traced back to a username, and of course the package name you're installing is the primary bit of the request…

This seems to get into the what the registry will do with this information. (The answer right now is "ignore it".) 

As for what we might do with it? What do we need to say beyond the tentative descriptions I've given below?

Something like? "We choose not to correlate the information in these headers with any authenticated accounts you're using."

ashleygwilliams replied:
yeah that last line sounds super good.

ashleygwilliams commented on doc/misc/npm-registry.md:
```diff
@@ -21,6 +21,21 @@ The registry URL used is determined by the scope of the package (see
+  organization.
+* `Npm-In-CI` – Set to "true" if npm believes this install is running in a
+  continous integration environment, "false" otherwise. This is detected by
+  looking for the following environment variables: `CI`, `TDDIUM`,
```
maybe link to the code that does this? i imagine people will want to argue technically about how this is incomplete and/or not accurate so it may be helpful to tell them where a PR would be useful.

ashleygwilliams replied:
👍

iarna committed:
npm-registry-client@7.3.0

Credit: @iarna
PR-URL: https://github.com/npm/npm-registry-client/pull/129
PR-URL: https://github.com/npm/npm-registry-client/pull/145

iarna commented:
This has been merged into `release-next`.
